---
title: "Seurat scRNAseq analysis"
author: "Rachel Griffard"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries and data
```{r}
library(scrublet)
library(Seurat)
library(ggplot2)
library(SingleR)
library(dplyr)
library(celldex)
library(RColorBrewer)
library(SingleCellExperiment)
```



```{r read-10X-data}
adj.matrix <- Read10X("dat")
```


# Create Seurat object
```{r}
srt = CreateSeuratObject(adj.matrix, project = "pbmc10k")
srt

# remove adj.matrix
remove(adj.matrix)
```

# Basic QC and filtering
```{r}
meta <- srt@meta.data
dim(meta)

head(meta)
```

Below we can see the summaries for the # of UMI reads per cell (nCount_RNA) and the 
number of expressed genes per that cell (nFeature_RNA)
```{r}
summary(meta$nCount_RNA)
summary(meta$nFeature_RNA)
```

Next, we will create extra columns identifying mitochondrial genes (MT-) as well as ribosomal proteins
(RPS or RPL)
```{r}
srt[["percent.mt"]] <- PercentageFeatureSet(srt, pattern = "^MT-")
srt[["percent.rb"]] <- PercentageFeatureSet(srt, pattern = "^RP[SL]")
```

## Include doublet annotation from *scrublet*
```{r}
doublets <- read.table("scrublet_calls.tsv",header = F,row.names = 1)
colnames(doublets) <- c("Doublet_score","Is_doublet")
srt <- AddMetaData(srt,doublets)
head(srt[[]])
```


## Visualizations from metadata
```{r}
VlnPlot(srt, features = c("nFeature_RNA","nCount_RNA","percent.mt","percent.rb"),ncol = 4,pt.size = 0.1) & 
  theme(plot.title = element_text(size=10))
```


```{r}
FeatureScatter(srt, feature1 = "nCount_RNA", feature2 = "percent.mt")
```


```{r}
FeatureScatter(srt, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```


```{r}
FeatureScatter(srt, feature1 = "nCount_RNA", feature2 = "percent.rb")
```


```{r}
FeatureScatter(srt, feature1 = "percent.rb", feature2 = "percent.mt")
```


```{r}
FeatureScatter(srt, feature1 = "nFeature_RNA", feature2 = "Doublet_score")
```


## Create QC column in metadata
```{r}
srt[['QC']] <- ifelse(srt@meta.data$Is_doublet == 'True','Doublet','Pass')
srt[['QC']] <- ifelse(srt@meta.data$nFeature_RNA < 500 & srt@meta.data$QC == 'Pass','Low_nFeature',srt@meta.data$QC)
srt[['QC']] <- ifelse(srt@meta.data$nFeature_RNA < 500 & srt@meta.data$QC != 'Pass' & srt@meta.data$QC != 'Low_nFeature',paste('Low_nFeature',srt@meta.data$QC,sep = ','),srt@meta.data$QC)
srt[['QC']] <- ifelse(srt@meta.data$percent.mt > 15 & srt@meta.data$QC == 'Pass','High_MT',srt@meta.data$QC)
srt[['QC']] <- ifelse(srt@meta.data$nFeature_RNA < 500 & srt@meta.data$QC != 'Pass' & srt@meta.data$QC != 'High_MT',paste('High_MT',srt@meta.data$QC,sep = ','),srt@meta.data$QC)
table(srt[['QC']])
```

## Violin plot for those that passed QC
```{r}
VlnPlot(subset(srt, subset = QC == 'Pass'), 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.rb"), ncol = 4, pt.size = 0.1) & 
  theme(plot.title = element_text(size=10))
```

# Normalize & dimension reduction

Normalize based on sequencing depth. Scaled by 10,000 and log-transform values.
```{r}
srt <- NormalizeData(srt)
```

Find most variable genes for downstream analysis.
```{r}
srt <- FindVariableFeatures(srt, selection.method = "vst", nfeatures = 2000)
```

Show top ten most variable
```{r}
top10 <- head(VariableFeatures(srt), 10)
top10
```

Plot variable features with top ten labeled
```{r}
plot1 <- VariableFeaturePlot(srt)
LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
```

We can use the ScaleData function to convert the normalized gene expression from above into Z-scores to input into PCA
```{r}
all.genes <- rownames(srt)
srt <- ScaleData(srt, features = all.genes) # forces to do all genes, not only top variable genes
```

Run PCA
```{r}
srt <- RunPCA(srt, features = VariableFeatures(object = srt))
```

Check heatmaps of principal components
```{r}
DimHeatmap(srt, dims = 1:6, nfeatures = 20, cells = 500, balanced = T)
```

Visualize dimension reduction and check elbow plot
```{r}
DimPlot(srat, reduction = "pca")
ElbowPlot(srat)
```

Find neighbors for cluster resolution in plot
```{r}
srat <- FindNeighbors(srat, dims = 1:10) # set dims by ElbowPlot
```

Find clusters, usually 0.8
```{r}
srat <- FindClusters(srat, resolution = 0.5)
```

Create and plot UMAP
```{r}
srat <- RunUMAP(srat, dims = 1:10, verbose = F)
DimPlot(srat,label.size = 4,repel = T,label = T)
FeaturePlot(srat, features = c("LILRA4", "TPM2", "PPBP", "GP1BB")) # DC(1:2), platelets(3:4)
```

Compare to UMAP without those that did not pass QC
```{r}
DimPlot(srat,label.size = 4,repel = T,label = T)
```

Calculate cell cycle scores
```{r}
s.genes <- cc.genes.updated.2019$s.genes
g2m.genes <- cc.genes.updated.2019$g2m.genes

srat <- CellCycleScoring(srat, s.features = s.genes, g2m.features = g2m.genes)
table(srat[[]]$Phase)
```

Mitochondrial UMAP
```{r}
FeaturePlot(srat,features = "percent.mt",label.size = 4,repel = T,label = T) & 
  theme(plot.title = element_text(size=10))
```

Ribosomal protein UMAP
```{r}
FeaturePlot(srat,features = "percent.rb",label.size = 4,repel = T,label = T) & theme(plot.title = element_text(size=10))
```

Cell cycle score UMAP
```{r}
FeaturePlot(srat,features = c("S.Score","G2M.Score"),label.size = 4,repel = T,label = T) & 
  theme(plot.title = element_text(size=10))
```

## 

